lib/db/set_test.go

```diff
@@ -892,6 +892,57 @@ func TestWithHaveSequence(t *testing.T) {
 	})
 }
 
+func TestIssue4925(t *testing.T) {
+	ldb := db.OpenMemory()
+
+	folder := "test)"
+	s := db.NewFileSet(folder, fs.NewFilesystem(fs.FilesystemTypeBasic, "."), ldb)
+
+	localHave := fileList{
+		protocol.FileInfo{Name: "dir"},
+		protocol.FileInfo{Name: "dir.file"},
+		protocol.FileInfo{Name: "dir/file"},
+	}
+
+	replace(s, protocol.LocalDeviceID, localHave)
+
+	i := 0
+	s.WithPrefixedHaveTruncated(protocol.LocalDeviceID, "dir", func(fi db.FileIntf) bool {
+		i++
+		return true
+	})
+	if i != 2 {
+		t.Errorf(`Expected 2, got %v local items below "dir"`, i)
+	}
+
+	i = 0
+	s.WithPrefixedHaveTruncated(protocol.LocalDeviceID, "dir/", func(fi db.FileIntf) bool {
+		i++
+		return true
+	})
+	if i != 1 {
+		t.Errorf(`Expected 1, got %v local items below "dir/"`, i)
+	}
+
+	i = 0
+	s.WithPrefixedHaveTruncated(protocol.LocalDeviceID, "dir", func(fi db.FileIntf) bool {
+		i++
+		return true
+	})
+	if i != 2 {
+		t.Errorf(`Expected 2, got %v global items below "dir"`, i)
+	}
+
+	i = 0
+	s.WithPrefixedGlobalTruncated("dir/", func(fi db.FileIntf) bool {
+		i++
+		return true
+	})
+	if i != 1 {
+		t.Errorf(`Expected 1, got %v global items below "dir/"`, i)
+	}
+}
```

Comment:
Just out of curiosity, was this failing previously on second time iterating? If yes, that would mean the iterators are somehow stateful? If not, I am confused why you do the same thing twice.